Classes/common/HTTP/CDTURLSession.m

```diff
@@ -56,22 +70,59 @@ - (instancetype)initWithDelegate:(NSObject<NSURLSessionDelegate> *)delegate
-                                                        NSURLResponse *response,
-                                                        NSError *error))completionHandler
+                              taskDelegate:(NSObject<CDTURLSessionTaskDelegate> *)taskDelegate
 {
-    CDTURLSessionTask *task = [[CDTURLSessionTask alloc] initWithSession:self.session
+    CDTURLSessionTask *task = [[CDTURLSessionTask alloc] initWithSession:self
                                                                  request:request
                                                             interceptors:self.interceptors];
-    __weak CDTURLSession *weakSelf = self;
-    task.completionHandler = ^void(NSData *data, NSURLResponse *response, NSError *error) {
-        __strong CDTURLSession *strongSelf = weakSelf;
-        if (strongSelf && completionHandler) {
-            data = [NSData dataWithData:data];
-            MYOnThread(strongSelf.thread, ^{ completionHandler(data, response, error); });
-        }
-    };
+    task.delegate = taskDelegate;
     return task;
 }
 
+- (NSURLSessionDataTask *)createDataTaskWithRequest:(NSURLRequest *)request
+                                 associatedWithTask:(CDTURLSessionTask *)task
+
+{
+    NSURLSessionDataTask *nsURLSessionTask = [self.session dataTaskWithRequest:request];
+    [_taskMap setObject:task forKey:[NSNumber numberWithInteger:nsURLSessionTask.taskIdentifier]];
+    return nsURLSessionTask;
+}
+
+- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
+{
+    __strong CDTURLSessionTask *cdtURLSessionTask = [self getSessionTaskForId:dataTask.taskIdentifier];
+    data = [NSData dataWithData:data];
+    MYOnThread(self.thread, ^{
+        [cdtURLSessionTask.delegate handleData:data];
+    });
+
+}
+
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+    __strong CDTURLSessionTask *cdtURLSessionTask = [self getSessionTaskForId:task.taskIdentifier];
+    if (error && cdtURLSessionTask) {
+        [cdtURLSessionTask processError:error onThread:self.thread];
+    }
+}
+
+- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveResponse:(NSURLResponse *)response completionHandler:(void (^)(NSURLSessionResponseDisposition))completionHandler
+{
+    __strong CDTURLSessionTask *cdtURLSessionTask = [self getSessionTaskForId:dataTask.taskIdentifier];
+    if (cdtURLSessionTask) {
+        [cdtURLSessionTask processResponse:response onThread:self.thread];
+    }
+    completionHandler(NSURLSessionResponseAllow);
+}
+
+- (CDTURLSessionTask *)getSessionTaskForId:(NSUInteger)identifier
+{
+    return [_taskMap objectForKey:[NSNumber numberWithInteger:identifier]];
+}
+
+- (void) disassociateTask:(nonnull NSURLSessionDataTask *)task
+{
+    [_taskMap removeObjectForKey:[NSNumber numberWithInteger:task.taskIdentifier]];
```

Comment:
Since this can be called from `dealloc` in the `CDTURLSessionTask` shouldn't this also cancel the inflight task to save network traffic since the result will be discarded?